pages/api/applications/[applicationId]/stages/index.ts

```diff
@@ -1,5 +1,146 @@
+    message: 'Application stage emoji unicode hex is invalid.',
+  },
+});
+
+async function handler(
+  uid: string,
+  req: NextApiRequest,
+  res: NextApiResponse<ApiResponse<ApplicationStageData | ApplicationDataWithStagesOnly>>,
+) {
+  switch (req.method) {
+    case HttpMethod.POST:
+      return handlePost(uid, req, res);
+    default:
+      return rejectHttpMethod(res, req.method);
+  }
+}
+
+async function handlePost(uid: string, req: NextApiRequest, res: NextApiResponse<ApiResponse<ApplicationStageData>>) {
+  const paramValidationError = validatePostRequest(req);
+  if (paramValidationError !== null) {
+    return res.status(HttpStatus.BAD_REQUEST).json(createJsonResponse({}, messages[paramValidationError]));
+  }
+
+  const application = await getApplication(uid, Number(req.query.applicationId));
+  if (application === null) {
+    return res.status(HttpStatus.NOT_FOUND).json(createJsonResponse({}, messages[MessageType.APPLICATION_NOT_FOUND]));
+  }
+
+  const { type, date, emojiUnicodeHex, remark } = req.body;
+  // Empty or undefined emojis and remarks will be set to null.
+  const parameters: ApplicationStagePostData = {
+    type,
+    date: new Date(date),
+    emojiUnicodeHex: emojiUnicodeHex || null,
+    remark: remark || null,
+  };
+
+  const newStage: ApplicationStageData = await prisma.applicationStage.create({
+    data: { ...parameters, applicationId: application.id },
+    select: { applicationId: true, id: true, type: true, date: true, emojiUnicodeHex: true, remark: true },
+  });
+
+  return res.status(HttpStatus.CREATED).json(createJsonResponse(newStage, messages[MessageType.CREATED]));
+}
+
+function validatePathParameters(req: NextApiRequest): Nullable<MessageType> {
+  return !canBecomeInteger(req.query.applicationId) ? MessageType.INVALID_APPLICATION_ID : null;
 }
+
+function validatePostRequest(req: NextApiRequest): Nullable<MessageType> {
+  const pathParameterError = validatePathParameters(req);
+  if (pathParameterError !== null) {
+    return pathParameterError;
+  }
+
+  const { type, date, emojiUnicodeHex, remark } = req.body;
+
+  if (typeof type !== 'string' || !(type in ApplicationStageType)) {
+    return MessageType.INVALID_TYPE;
+  }
+
+  if (date === null || date === undefined || !isValidDate(date + '')) {
+    return MessageType.INVALID_DATE;
+  }
+
+  const isValidRemark = remark === null || remark === undefined || typeof remark === 'string';
+  if (!isValidRemark) {
+    return MessageType.INVALID_REMARK;
+  }
+
+  const isValidEmojiUnicodeHex =
+    emojiUnicodeHex === null ||
+    emojiUnicodeHex === undefined ||
+    (typeof emojiUnicodeHex === 'string' && isValidHex(emojiUnicodeHex));
+  if (!isValidEmojiUnicodeHex) {
+    return MessageType.INVALID_EMOJI_UNICODE_HEX;
+  }
+
+  return null;
+}
+
+/** Finds application if present for the user, else returns null. */
+async function getApplication(uid: string, applicationId: number): Promise<Nullable<ApplicationDataWithStagesOnly>> {
```

Comment:
We are only checking if the application is found. Do we need this method and type?

I don't think we need this method.